### DISCUSSION TITLE
feature request: Adding new menu item in reader context menu

### DISCUSSION THREAD

**Discussion by vkiranmaniya** [edrlab/thorium-reader#1773] [Developer corner]
I have added the context menu item for the reader window. There already one `Inspect Element` menu entry for dev, I want another entry for a production build. What I'm trying to do is I want context menu item `Add to Bookmark` and in the click, handler passes the IPC call to renderer process to call the add to bookmark action. But the menu item `Add to bookmark` won't appear. 
I' using the latest `develop` branch.

**Comment by panaC** [comment]
i don't understand precisely what do you looking for.

But if you want to create a new menu in the top bar, our codebase is here : https://github.com/edrlab/thorium-reader/blob/694b9c63f22db039a1af055b637ea75d2b76ad99/src/main/redux/sagas/win/browserWindow/createReaderWindow.ts#L49

and then you can dispatch a main redux action on click.

it should work in prod env.

**Comment by vkiranmaniya** [comment]
Here is the code, The issue is when i right click on the middle of window, Add to bookmark button doesn't appear and it's showing else where. when i right click on setting icon, it's visible in context menu.
```js
const wc = readerWindow.webContents;
    wc.on("context-menu", (_ev, params) => {
        const { x, y } = params;
        const openDevToolsAndInspect = () => {
            const devToolsOpened = () => {
                wc.off("devtools-opened", devToolsOpened);
                wc.inspectElement(x, y);

                setTimeout(() => {
                    if (wc.isDevToolsOpened() && wc.devToolsWebContents) {
                        wc.devToolsWebContents.focus();
                    }
                }, 500);
            };
            wc.on("devtools-opened", devToolsOpened);
            wc.openDevTools({ activate: true, mode: "detach" });
        };
        Menu.buildFromTemplate([{
            label: "Add to Bookmark",
            click: () => {
                console.log("Add to Bookmark Option Clicked");
                // TODO implement the add bookmark function by PRC call
            },
        }, IS_DEV ? {
            click: () => {
                const wasOpened = wc.isDevToolsOpened();
                if (!wasOpened) {
                    openDevToolsAndInspect();
                } else {
                    if (!wc.isDevToolsFocused()) {
                        // wc.toggleDevTools();
                        wc.closeDevTools();

                        setImmediate(() => {
                            openDevToolsAndInspect();
                        });
                    } else {
                        wc.inspectElement(x, y);
                    }
                }
            },
            label: "Inspect element",
        } : undefined]).popup({window: readerWindow});
    });
```

Also, How to pass the RPC call to add to bookmark? Here is what i decided to go with.
onclick of context menu item, main process sends RPC event to the reader window and from the renderer process we can call add to bookmark function. If there is better way, please suggest.

**Comment by danielweck** [comment]
Hello, the "central" part of the reader window is not managed by Thorium's code, it is implemented by the "navigator" component which handles the complex layout and rendering logic of publication resources (in other words, the application user interface is the "chrome" that surrounds this central area).

Application developers are not expected to modify / tweak code in the "navigator" component, so ideally / ultimately this component should provide hooks / extension points so that additional context menus / etc. can be plugged in. Right now, this service / API does not exist, I am afraid.

Could you please explain what your use case is? What feature do you need the menu items to expose? How much interaction with the rendered DOM of publication HTML documents do you need? (this last question is the main issue, because publication resources are embedded in a strict sandbox, because of security considerations when loading foreign / untrusted content)

**Comment by danielweck** [comment]
PS: the "add bookmark" feature that already exists in Thorium takes into account the "current reading location" and even the currently-selected text in the webview that hosts publication HTML documents. Users can currently click on the bookmark icon-button (top toolbar), of use the CTRL-B keyboard shortcut (see https://github.com/edrlab/thorium-reader/wiki/Keyboard-shortcuts ).
I can see how a context menu may provide an additional useful way to invoke the "add bookmark" command, but as explained above, there is currently no API in the "navigator" component to expose this kind of service.

**Comment by vkiranmaniya** [comment]
@danielweck Yeh your answer makes sense, I also figured that out because the context menu item doesn't appear over the reader view only. My use case is when user selects text and right clicks, I want to provide add to bookmark action in context menu.

**Comment by vkiranmaniya** [comment]
Will the readium.js support that hook in the near future? I'm heavily relying upon that.

**Comment by danielweck** [comment]
There are no concrete plans, I am afraid. We have several other high development priorities at the moment, and only a small engineering team with finite time.
The open-source r2-navigator-js component contains code that displays the debug (“inspect element”) pop up context menu, perhaps you could take a look there, and figure out how to add an IPC event to communicate Thorium’s reader BrowserWindow (i.e. IPC messaging between the web view process of the publication HTML document, and the Electron renderer process of the main BowserWindow)

**Comment by vkiranmaniya** [comment]
Thanks for the hint. It helps a lot.

**Comment by panaC** [comment]
https://github.com/edrlab/thorium-reader/blob/233283ed85d3f3f120cabee58e6200265d88a365/src/renderer/reader/components/Reader.tsx#L827

https://github.com/edrlab/thorium-reader/blob/321c18b97bdee43f3619de085d2a733f5ad6ccc2/src/renderer/common/redux/api/api.ts#L19-L29

put this on the click function : `apiDispatch(store.dispatch)()("reader/addBookmark")(pubId, locator, name)`

**Comment by panaC** [comment]
to access to the reader reduxState  follow the 'session' key in the main reduxState.
https://github.com/edrlab/thorium-reader/blob/233283ed85d3f3f120cabee58e6200265d88a365/src/main/redux/reducers/win/session/reader.ts#L83

**Comment by danielweck** [comment]
Moving this to a developer discussion

**Comment by danielweck** [comment]
Related: https://github.com/edrlab/thorium-reader/discussions/2968

**Comment by danielweck** [comment]
The main "blocker" for this functional evolution of Thorium's Graphical User Interface is not technical, it is more about the fact that the right-click context menu was originally introduced only in debug mode (only activated in test builds) in order to inspect the current DOM element. There is a similar UX evolution problem with the system-level menubar (i.e. depending on the operating system, either displayed at the top of the application window(s) or at the top of the screen): this is technically feasible but the user experience (UX) wasn't originally designed with a menubar in mind.
Purely from a technical perspective, we need to figure out the "inversion of control" from the HTML document which the user interacts with, to the application host which controls what menu to display at the mouse / pointing device cursor location.
